src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/RuntimeBinderExtensions.cs

```diff
@@ -278,11 +278,36 @@ private static bool IsTypeParameterEquivalentToTypeInst(this Type typeParam, Typ
 
         public static bool HasSameMetadataDefinitionAs(this MemberInfo mi1, MemberInfo mi2)
         {
+            if (s_lazyHasSameMetadataDefinitionAsApi == null)
+            {
+                MethodInfo apiMethod = typeof(MemberInfo).GetMethod(
```

Comment:
I think this should be added to the self-modifying s_MemberEquivalence above. It woudl be just another `try/catch` before we check for the MetadataToken.

Note that the pattern not only checks for the presence of the API, but also that it does not throw right away. We had major issues with API being discovered and then throwing something like "Not supported" or "Method not avaialble" - I do not remember. 
The pattern is to 
1 - probe for the API
2 - try using the API for the given args
3 - if API worked without exceptions, self-modify the func to  the new implementation so we are no longer wasting time on various conditional checks.
4 - return the result.